src/Combinatorics/EnumerativeCombinatorics/weak_compositions.jl

```diff
@@ -0,0 +1,182 @@
+################################################################################
+#
+#  Constructors and basic functionality
+#
+################################################################################
+
+data(C::WeakComposition) = C.c
+
+@doc raw"""
+    weak_composition(parts::Vector{T}; check::Bool = true) where T <: IntegerUnion
+
+Return the weak composition given by the integer sequence `parts` as an object of type
+`WeakComposition{T}`.
+
+If `check` is `true` (default), it is checked whether the given sequence defines
+a weak composition, that is, whether all elements of `parts` are non-negative.
+
+# Examples
+```jldoctest
+julia> W = weak_composition([6, 0, 2, 3]) # the weak composition 6, 0, 2, 3 of 11
+[6, 0, 2, 3]
+
+julia> W = weak_composition(Int8[6, 0, 2, 3]) # save the elements in 8-bit integers
+Int8[6, 0, 2, 3]
+```
+"""
+function weak_composition(parts::Vector{T}; check::Bool = true) where {T <: IntegerUnion}
+  if check
+    @req all(x -> x >= 0, parts) "The integers must be non-negative"
+  end
+  return WeakComposition{T}(parts)
+end
+
+function Base.show(io::IO, ::MIME"text/plain", C::WeakComposition)
+  print(io, data(C))
+end
+
+################################################################################
+#
+#  Array-like functionality
+#
+################################################################################
+
+function Base.size(C::WeakComposition)
+  return size(data(C))
+end
+
+function Base.length(C::WeakComposition)
+  return length(data(C))
+end
+
+function Base.getindex(C::WeakComposition, i::IntegerUnion)
+  return getindex(data(C), Int(i))
+end
+
+function Base.setindex!(C::WeakComposition, x::IntegerUnion, i::IntegerUnion)
+  return setindex!(data(C), x, Int(i))
+end
+
+function Base.copy(C::WeakComposition)
+  return weak_composition(copy(data(C)))
+end
```

Comment:
The `copy` is here because otherwise `copy(::WeakComposition)` would produce a `Vector`. `deepcopy` seems to work fine presumably because the type `WeakComposition` has just one field.

Reply:
Yeah, for basic recursive deepcopying there is a fallback. For most of our types we have to add something on our own to not create copies of the parents.

Reply:
Sure. These don't have parents.